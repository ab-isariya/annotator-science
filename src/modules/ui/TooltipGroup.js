import {Children, cloneElement} from 'react';
import PropTypes from 'prop-types';

import Tippy, {useSingleton} from '@tippyjs/react';

//NOTE(Rejon): This component mimic's Tippy's Singleton example: https://github.com/atomiks/tippyjs/blob/master/website/src/components/examples/Singleton.js
const TooltipGroup = ({children, delay, size, dark, theme, transition}) => {
  const [source, target] = useSingleton();

  let initialDelay = delay || 500;
  const tooltipDelay = transition ? [100, initialDelay] : initialDelay;

  //Should build out a string like 'annotate-small' or 'annotate-dark-large'.
  //NOTE(Rejon): See theme name definitions in index.css.
  //NOTE(Rejon): We have to apply style here as well because our singleton Tippy instance won't render
  //				our styles if we don't provide them to the top level parent tooltip.
  const styleTheme = theme || `annotate${dark ? '-dark' : ''}-${size}`;

  return (
    <>
      <Tippy
        singleton={source}
        delay={tooltipDelay}
        theme={styleTheme}
        moveTransition={
          transition ? 'transform 0.4s cubic-bezier(0.22, 1, 0.36, 1)' : ''
        }
      />
      {Children.map(children, (child) =>
        cloneElement(child, {singleton: target})
      )}
    </>
  );
};

TooltipGroup.defaultProps = {
  size: 'medium'
};

TooltipGroup.propTypes = {
  //Children/Targets of the TooltipGroup
  children: PropTypes.node.isRequired,
  //Millisecond delay of the tooltip leaving after moving our mouse away
  delay: PropTypes.number,
  //If the tooltips should animate between positions or not.
  transition: PropTypes.bool,
  //Tooltip theme size (large, medium, or small)
  //medium is default
  //NOTE(Rejon): See index.css for theme size definitions
  size: PropTypes.oneOf(['medium', 'small', 'large']),
  //Tooltip dark theme, if this prop is on the tooltip it will use
  //an alternate dark theme defined in index.css
  dark: PropTypes.any,
  //Custom theme name incase we want to use an existing tippy theme instead.
  theme: PropTypes.string
};

export default TooltipGroup;
