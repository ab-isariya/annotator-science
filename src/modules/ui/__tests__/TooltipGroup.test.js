import ReactDOM from 'react-dom';
import TooltipGroup from '../TooltipGroup';
import renderer from 'react-test-renderer';

beforeAll(() => {
  ReactDOM.createPortal = jest.fn((element, node) => {
    return element;
  });
});

describe('<TooltipGroup/>', () => {
  it('Renders without crashing', () => {
    const comp = renderer
      .create(
        <TooltipGroup>
          <p>Hello</p>
        </TooltipGroup>
      )
      .toJSON();

    expect(comp).toMatchSnapshot();
  });

  it('Renders with children', () => {
    const comp = renderer
      .create(
        <TooltipGroup>
          <p>Hello</p>
        </TooltipGroup>
      )
      .toJSON();

    expect(comp).toMatchSnapshot();
  });

  it('Renders with dark and a set size', () => {
    const comp_dark = renderer
      .create(
        <TooltipGroup dark>
          <p>Hello</p>
        </TooltipGroup>
      )
      .toJSON();

    expect(comp_dark).toMatchSnapshot();

    const comp_size = renderer
      .create(
        <TooltipGroup size={'large'}>
          <p>Hello</p>
        </TooltipGroup>
      )
      .toJSON();

    expect(comp_size).toMatchSnapshot();
  });

  it('Renders with delay and transition', () => {
    const comp_delay = renderer
      .create(
        <TooltipGroup delay={1000}>
          <p>Hello</p>
        </TooltipGroup>
      )
      .toJSON();

    expect(comp_delay).toMatchSnapshot();

    const comp_transition = renderer
      .create(
        <TooltipGroup transition>
          <p>Hello</p>
        </TooltipGroup>
      )
      .toJSON();

    expect(comp_transition).toMatchSnapshot();
  });

  it('Renders with a custom theme', () => {
    const comp = renderer
      .create(
        <TooltipGroup theme="light">
          <p>Hello</p>
          <p>World</p>
        </TooltipGroup>
      )
      .toJSON();

    expect(comp).toMatchSnapshot();
  });
});
