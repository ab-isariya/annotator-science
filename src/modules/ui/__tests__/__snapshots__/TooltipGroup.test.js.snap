// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TooltipGroup/> Renders with a custom theme 1`] = `
Array [
  <p
    singleton={
      Object {
        "cleanup": [Function],
        "hook": [Function],
      }
    }
  >
    Hello
  </p>,
  <p
    singleton={
      Object {
        "cleanup": [Function],
        "hook": [Function],
      }
    }
  >
    World
  </p>,
]
`;

exports[`<TooltipGroup/> Renders with children 1`] = `
<p
  singleton={
    Object {
      "cleanup": [Function],
      "hook": [Function],
    }
  }
>
  Hello
</p>
`;

exports[`<TooltipGroup/> Renders with dark and a set size 1`] = `
<p
  singleton={
    Object {
      "cleanup": [Function],
      "hook": [Function],
    }
  }
>
  Hello
</p>
`;

exports[`<TooltipGroup/> Renders with dark and a set size 2`] = `
<p
  singleton={
    Object {
      "cleanup": [Function],
      "hook": [Function],
    }
  }
>
  Hello
</p>
`;

exports[`<TooltipGroup/> Renders with delay and transition 1`] = `
<p
  singleton={
    Object {
      "cleanup": [Function],
      "hook": [Function],
    }
  }
>
  Hello
</p>
`;

exports[`<TooltipGroup/> Renders with delay and transition 2`] = `
<p
  singleton={
    Object {
      "cleanup": [Function],
      "hook": [Function],
    }
  }
>
  Hello
</p>
`;

exports[`<TooltipGroup/> Renders without crashing 1`] = `
<p
  singleton={
    Object {
      "cleanup": [Function],
      "hook": [Function],
    }
  }
>
  Hello
</p>
`;
